src/file_pipe_log/log_file.rs

```diff
@@ -6,35 +6,114 @@ use std::io::{Read, Seek, SeekFrom, Write};
 use std::sync::Arc;
 
 use crate::metrics::*;
-use crate::pipe_log::FileBlockHandle;
+use crate::pipe_log::{FileBlockHandle, FileSeq};
 use crate::{Error, Result};
 
-use super::format::LogFileHeader;
+use super::format::{LogFileHeader, Version};
 use crate::env::{FileSystem, Handle, WriteExt};
 
 /// Maximum number of bytes to allocate ahead.
 const FILE_ALLOCATE_SIZE: usize = 2 * 1024 * 1024;
 
+/// Combination of `[Handle]` and `[Version]`, specifying a handler of a file.
+#[derive(Clone, Debug)]
+pub struct FileHandler<F: FileSystem> {
+    pub seq: FileSeq,
+    pub handle: Arc<F::Handle>,
+    pub version: Option<Version>,
+}
+
+/// Function for reading the header of the log file, and return a
+/// `[LogFileHeader]`.
+///
+/// Attention please, to avoid to move the offset of the given `[handle]`, we
+/// use a copy of the `[handle]` to parse the header of the given file.
+pub(super) fn build_file_header<F: FileSystem>(
+    system: &F,
+    handle: Arc<F::Handle>,
+    expected_reader: Option<&mut F::Reader>,
+) -> Result<LogFileHeader> {
+    let file_size: usize = match handle.file_size() {
+        Ok(size) => size,
+        Err(_) => {
+            return Err(Error::Corruption("Corrupted file!".to_owned())); // invalid file
+        }
+    };
+    // [1] If the file was a new file, we just return the default `LogFileHeader`.
+    if file_size == 0 {
+        return Ok(LogFileHeader::default());
+    }
+    // [2] If the length lessed than the standard `LogFileHeader::len()`.
+    let header_len = LogFileHeader::len();
+    if file_size < header_len {
+        return Err(Error::Corruption("Invalid header of LogFile!".to_owned()));
+    }
+    // [3] Parse the header of the file.
+    let parse_file_header = |handle: Arc<F::Handle>, header_len: usize| -> Result<LogFileHeader> {
```

Comment:
BTW, why use a closure here?

Reply:
Oh, it's just a personal code-sytle. As the capture of variable are implemented as `moved`,  it actually made no difference to the direct implementation without the closure. I just wanna to integrate the implementation of `parsing the header` here.

Reply:
You can define a function inside a function. Closure is only needed when we need to dynamically capture a variable.

Reply:
Got it.